### DISCUSSION TITLE
Testbed Automation

### DISCUSSION THREAD

**Discussion by susanta-mukhopadhyay** [9tigerio/db2rest#223] [Ideas]
Do we have any plan for providing automated testbed for the APIs we are generating? It may be simple newman script with option to feed data, which can be integrated with CI pipeline easily.

**Comment by kdhrubo** [comment]
It will be good if you can elaborate more on this. How do you expect this to work end to end ?

Open API specification will be available in next release. Is it possible to generate Newman script from it ?

**Reply by susanta-mukhopadhyay** [reply]
Newman is command line test collection runner for postman. I think test collections can not be generated from Open API directly. We may think of generating some test collection for testing the generated API. We also have similar API test tool like bruno. I shall check further about how we can do it easily in our current framework.

**Reply by kdhrubo** [reply]
May be we can provide another end point that generates these test scripts to download? 

or a cli ? more work i believe

**Reply by thadguidry** [reply]
I'd recommend Playwright because it can handle [a multitude of different contexts](https://playwright.dev/docs/api-testing) out-of-the-box.  I only used it once for light HTML browser testing because [it can generate tests easily in UI](https://playwright.dev/docs/codegen), but have heard that it's API testing is also very good because of the [Testing options](https://playwright.dev/docs/test-configuration) and [advanced configuration](https://playwright.dev/docs/test-configuration#advanced-configuration) it offers.

**Reply by kdhrubo** [reply]
Don't see any scripting options that playwright can drive.

For Newman , it can drive any postman collection from pipeline..

**Reply by susanta-mukhopadhyay** [reply]
I have not used Playwright, but another option is bruno. It provides option to create test collection script as well and slightly better in performance. https://github.com/usebruno/bruno .

**Reply by kdhrubo** [reply]
Also httpie.

**Comment by kdhrubo** [comment]
I believe we can provide a CLI that should create these scripts and keep it away from main db2rest

**Comment by kdhrubo** [comment]
How do we know what to test besides the generic crud ?

**Reply by susanta-mukhopadhyay** [reply]
Good point. Can we provide some interface or command to generate generic test script with option to edit and provide test data? Based on the column names we can also think of providing separate json file to be parsed and picked up by the test scripts.

**Comment by susanta-mukhopadhyay** [comment]
While using it in development, end user will anyway need test script to be integrated in their pipeline. If we can provide some automation it will be value addition for this product.
